numpy/core/src/multiarray/common.h

```diff
@@ -158,9 +162,21 @@ check_and_adjust_axis(int *axis, int ndim)
             }
         }
 
-        PyErr_Format(AxisError_cls,
-                     "axis %d is out of bounds for array of dimension %d",
-                     *axis, ndim);
+        /* build arguments for the AxisError constructor */
+        args = Py_BuildValue("iiO", *axis, ndim, msg_prefix);
+        if(args == NULL) {
+            return -1;
+        }
+
+        /* Invoke the AxisError constructor */
+        exc = PyObject_CallObject(AxisError_cls, args);
```

Comment:
Implementing the string formatting in C seemed like way too much effort for little gain - no need for performance once something starts going wrong